blockchain/blockchain.go

```diff
@@ -240,9 +247,213 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
+	// TODO-Klaytn consider to check Trie contents optionally
+	// TODO-Klaytn consider to check storage trie also
+	dirty, err := bc.checkTrieContents(targetDB, srcCachedDB, rootHash)
+	if err != nil || len(dirty) > 0 {
+		logger.Error("copied stated is invalid", "err", err, "len(dirty)", len(dirty))
+		// TODO-Klaytn Remove new DB and log.Error
+		if err != nil {
+			return err
+		}
+
+		return errors.New("copied state is not same with origin")
+	}
+
+	// TODO-Klaytn Swap DB and Remove Old DB
+	bc.db.FinishStateMigration()
+	logger.Info("completed state migration")
+
+	return nil
+}
+
+func (bc *BlockChain) checkTrieContents(oldDB, newDB *statedb.Database, root common.Hash) ([]common.Address, error) {
+	oldTrie, err := statedb.NewSecureTrie(root, oldDB)
+	if err != nil {
+		return nil, err
+	}
+
+	newTrie, err := statedb.NewSecureTrie(root, newDB)
+	if err != nil {
+		return nil, err
+	}
+
+	diff, _ := statedb.NewDifferenceIterator(oldTrie.NodeIterator([]byte{}), newTrie.NodeIterator([]byte{}))
+	iter := statedb.NewIterator(diff)
+
+	var dirty []common.Address
+
+	for iter.Next() {
+		key := newTrie.GetKey(iter.Key)
+		if key == nil {
+			return nil, fmt.Errorf("no preimage found for hash %x", iter.Key)
+		}
+
+		dirty = append(dirty, common.BytesToAddress(key))
+	}
+
+	return dirty, nil
+}
+
+func (bc *BlockChain) checkRestartStateMigration() {
+	if bc.db.InMigration() {
+		number := bc.db.MigrationBlockNumber()
+
+		block := bc.GetBlockByNumber(number)
+		if block == nil {
+			logger.Error("failed to get migration block number", "blockNumber", number)
+			return
+		}
+
+		root := block.Root()
+		logger.Warn("Restart state trie migration", "blockNumber", number, "root", root.String())
+
+		go bc.migrateState(root)
+	}
+}
+
+func (bc *BlockChain) PrepareStateMigration() error {
+	if bc.db.InMigration() || bc.prepareStateMigration {
+		return errors.New("migration already started")
+	}
+
+	bc.prepareStateMigration = true
+	currentBlock := bc.CurrentBlock().NumberU64()
+	nextCommittedBlock := currentBlock + (DefaultBlockInterval - currentBlock%DefaultBlockInterval)
+	logger.Warn("Prepare state migration", "nextCommittedBlock", nextCommittedBlock)
+
+	return nil
+}
+
+func (bc *BlockChain) checkStartStateMigration(number uint64, root common.Hash) {
+	if bc.prepareStateMigration {
+		logger.Info("Start state migration", "block", number, "root", root)
+
+		if err := bc.StartStateMigration(number, root); err != nil {
+			logger.Error("Failed to start state migration", "err", err)
+		}
+
+		bc.prepareStateMigration = false
+	}
+}
+
+func (bc *BlockChain) StartStateMigration(number uint64, root common.Hash) error {
+	// TODO-Klaytn Add internal status check routine
+	if bc.db.InMigration() {
+		return errors.New("migration already started")
+	}
+
+	if err := bc.db.CreateMigrationDBAndSetStatus(number); err != nil {
+		return err
+	}
+
+	go bc.migrateState(root)
+
+	return nil
+}
+
+func (bc *BlockChain) StopStateMigration() error {
```

Comment:
This function should be called in `(bc *BlockChain) Stop()`. If not, Blockchain will not stop during migration because of `bc.wg.Wait()`.

Reply:
`(bc *BlockChain) Stop()` closes `<-bc.quit`. So `<-bc.quit` can stop the stateMigration.
So `(bc *BlockChain) Stop()` does not need to call `StopStateMigration()`.